Third Party Identity Management

juleengraham opened the pull request:
Implementation of UI that allows users to add and disconnect third party accounts to/from their Galaxy account.

juleengraham committed:
display connected external ids

disconnecting is not yet working

trying to determine why disconnect won't work

juleengraham committed:
display connected external ids

disconnecting is not yet working

trying to determine why disconnect won't work

juleengraham committed:
Implement External Id disconnect UI

add features to external identities

- allow display of email with each external identity
- check if user is disconnecting final identity

Latest changes

juleengraham committed:
user preferences

- add external id management to userpreferences vue file
- standardise capitalisation of titles (some had camel case, some sentence case, others followed neither rule)

nsoranzo commented on lib/galaxy/authnz/psa_authnz.py:
```diff
@@ -357,12 +357,12 @@ def disconnect(name=None, user=None, user_storage=None, strategy=None,
     disconnect pipeline, and that value will be returned as a result of calling the `do_disconnect` function.
     """
-    user_authnz = strategy.trans.sa_session.query(user_storage).filter(user_storage.table.c.user_id == user.id,
+    user_authnz = user_storage.sa_session.query(user_storage).filter(user_storage.table.c.user_id == user.id,
```
I'd assign `user_storage.sa_session` to a variable (e.g. `sa_session`) and use that in the 3 occurrences.

nsoranzo commented on lib/galaxy/config/sample/galaxy.yml.sample:
```diff
@@ -1465,18 +1465,12 @@ galaxy:
 
   # Enables and disables OpenID Connect (OIDC) support.
-  #enable_oidc: false
+  enable_oidc: true
```
I suppose you didn't mean to actually commit changes to this file?

martenson commented on client/galaxy/scripts/components/User/ExternalIdentities/ExternalIdentities.vue:
```diff
@@ -0,0 +1,394 @@
+};
+
+/*
+    data() {
```
Please drop the commented out section

martenson commented on client/galaxy/scripts/components/User/ExternalIdentities/ExternalIdentities.vue:
```diff
@@ -0,0 +1,394 @@
+            background: none;
+        }
+
+        // Undo butchering from base.css so side-by-side labels work
```
This comment indicates we should be fixing something in base.scss instead?

martenson commented on client/galaxy/scripts/components/User/ExternalIdentities/ExternalIdentities.vue:
```diff
@@ -0,0 +1,394 @@
+            svc.saveIdentity(idp)
+                .then(response => {
+                    if (response.data.redirect_uri) {
+                        //window.location.reload();
```
Please drop (and all other commented out code too).

martenson commented on client/galaxy/scripts/components/User/ExternalIdentities/ExternalIdKey.vue:
```diff
@@ -0,0 +1,47 @@
+                                aria-label="Disconnect External Identity"
+                                title="Disconnect External Identity"
+                            >
+                                <span>Disconnect External Identityyyyyyyyyy</span>
```
This does not look right.

mvdbeek replied:
That still doesn't look right ;)

martenson commented on client/galaxy/scripts/components/User/ExternalIdentities/ExternalIdentities.vue:
```diff
@@ -0,0 +1,394 @@
+                and password, or their Google account. Whichever method they use they will be assuming same Galaxy 
+                user account, hence having access to the same histories, workflows, datasets, libraries, etc.
+                
+                <br>See more information, including a list of supported identity providers, 
```
Could we please use e.g. paragraphs instead of breaks for better semantics?

martenson reviewed (commented):
Couple of first pass notes inline.

dannon committed:
Client format

PR updates

- Fix linting issues
- Disable third-party management if oidc login is not enabled
- Make sign on images (PR 8107) clickable

mvdbeek commented on lib/galaxy/authnz/psa_authnz.py:
```diff
@@ -404,12 +404,14 @@ def disconnect(name=None, user=None, user_storage=None, strategy=None,
+
+    sa_session = user_storage.sa_session
+    user_authnz = sa_session.query(user_storage).filter(user_storage.table.c.user_id == user.id,
                                                                        user_storage.table.c.provider == name).first()
```
There's a linting error in https://app.circleci.com/jobs/github/galaxyproject/galaxy/56933,
you'll want to put the 2nd argument to start at the same column as the first argument:
```suggestion
                                                        user_storage.table.c.provider == name).first()
```
(this is kind-of approximate in the github UI)

mvdbeek commented:
There are also some jslint errors here:
```
yarn run v1.19.1
$ eslint -c .eslintrc.js galaxy/scripts --ext .js,.vue

/home/circleci/repo/client/galaxy/scripts/components/User/ExternalIdentities/ExternalIdentities.vue
   81:9   error  The "ExternalIdKey" component has been registered but not used  vue/no-unused-components
  179:19  error  'currentUrl' is assigned a value but never used                 no-unused-vars

/home/circleci/repo/client/galaxy/scripts/components/User/ExternalIdentities/service.js
   7:10  error  'getCurrentUser' is defined but never used     no-unused-vars
  22:5   error  'identityProviders' is defined but never used  no-unused-vars

✖ 4 problems (4 errors, 0 warnings)
```

martenson commented:
@juleengraham could you please make this hidden behind the `enable_oidc` configuration option so it is not displayed when not relevant?

juleengraham committed:
Implement Login via CILogon

juleengraham commented:
Resolve conflicts with upstream

juleengraham committed:
Logouts After Final Disconnect

- Logs user out of account when they disconnect their final external identity.

jdavcs commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -230,3 +247,12 @@ def _get_verify_param(self):
+        temp_username = email.split('@')[0] #username created from username portion of email
+        count = 0
+
+        while (trans.sa_session.query(trans.app.model.User).filter_by(username=(temp_username+str(count))).first()):
```
this doesn't check for existing username without the appended count - i.e., you start with appending zero (i.e., before probing for `jack0`, you need to check the availability of `jack`, right?)

jdavcs commented on lib/galaxy/config/sample/galaxy.yml.sample:
```diff
@@ -1514,14 +1514,11 @@ galaxy:
-  # The value of this option will be resolved with respect to
-  # <config_dir>.
-  #oidc_config_file: oidc_config.xml
+  oidc_config_file: config/oidc_config.xml
```
1) Both oidc config properties are resolved w.r.t. `config_dir` automatically, so the original values are correct.
2) As per @nsoranzo , you should not be modifying this file directly. Any changes should be made to `webapps/galaxy/config_schema.yml`, after which you generate it with `make config-rebuild`.

jdavcs commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -230,3 +247,12 @@ def _get_verify_param(self):
+        count = 0
+
+        while (trans.sa_session.query(trans.app.model.User).filter_by(username=(temp_username+str(count))).first()):
+        #if username already exists in database, append integer and iterate until unique username found
```
The number of near-identical usernames could be high in some cases ("example#", "test#", "john.smith#"...). Issuing sequential queries to the db to retrieve each one would be unnecessarily expensive. I'm not familiar with the context, so it could be a non-issue. Otherwise, you could maybe retrieve all partial matches in one query and generate the next username+count based on that result?

juleengraham committed:
Creating b-modal

juleengraham committed:
Attempt to avvoid CORS error

Fix idp vue warning + small changes

juleengraham committed:
Remove prints and console logs

nsoranzo commented on lib/galaxy/authnz/managers.py:
```diff
@@ -294,8 +304,11 @@ def logout(self, provider, trans, post_logout_redirect_url=None):
     def disconnect(self, provider, trans, disconnect_redirect_url=None):
+=======
+    def disconnect(self, provider, trans, disconnect_redirect_url=None, idphint=None):
+>>>>>>> 1683238e71... Implement Login via CILogon
```
There is a bad merge here.

nsoranzo commented on lib/galaxy/config/sample/galaxy.yml.sample:
```diff
@@ -1,2014 +0,0 @@
-# Galaxy is configured by default to be usable in a single-user development
```
This file should not be removed! Bad merge/rebase?

juleengraham committed:
Remove unnecessary files

juleengraham committed:
Integrate both CILogon and Custos using one list of idps

- run prettier
- remove commented code
- other minor changes (remove authenticationDuplicate exception class)

juleengraham committed:
Update custos api endpoints

juleengraham committed:
Add email for CILogon and Custos connected identities

dannon committed:
Drop unused requests import.

dannon committed:
Swap from vue-selectize to vue-multiselect, which we've chosen for use elsewhere in Galaxy as the likely select2 replacement

dannon committed:
Disable custos/cilogon submit button until there is an entity selected

dannon committed:
Login cleanup/formatting.

dannon committed:
Incremental work fixing up the formatting when using institutional login

dannon committed:
Use form group for consistent spacing/messaging.  Fix overflow styling
with vue-multiselect box.

dannon committed:
Client format

dannon committed:
Cleanup custos_authnz...

dannon committed:
Drop unused extra query

dannon committed:
Fix custos authnz bug where b64encode was getting a dumb string

dannon committed:
Drop directionality specification on cilogon institution list

dannon committed:
Fix python linting.

dannon committed:
Fix custos auth bug introduced in 8d9a56f68b77bf7b37224cc5cd7f081adcec8701

dannon committed:
Add fix for mocked custos authnz test

dannon committed:
Revert change to custos_authnz callback return signature

dannon committed:
Fix generation of authorization url?

dannon committed:
Set cilogon/custos default idp_hint param correctly.

dannon committed:
Fix callback signature/special casing for custos/cilogon

dannon committed:
Update default custos idp in test.

dannon committed:
Merge remote-tracking branch 'upstream/dev' into ext-id

dannon committed:
Fixes post-merge to handle list being an object now

dannon committed:
Fix linting issues with hasOwnProperty

dannon committed:
Drop debug cruft

afgane reviewed (approved):
Functionally, this is working really nice!

dannon committed:
Typo that was once fixed, but lost in a rebase...

dannon committed:
Fix oidc_idp handling to account for changes, drop some unused functionality.

dannon committed:
Fix logout after disconnection; gotta actually call this.

dannon committed:
Fix duplicate ID causing error with showing the wrong modal

dannon committed:
Refactor display if disconnect buttons; fix up styling.  Fix a typo.

dannon committed:
rephrase last remaining identity deletion protocol

dannon committed:
Text formatting fail/ rerun prettier.
